providers/flagd/pkg/provider.go

```diff
@@ -0,0 +1,170 @@
+package flagd
+
+import (
+	"github.com/open-feature/golang-sdk-contrib/providers/flagd/pkg/service"
+	GRPCService "github.com/open-feature/golang-sdk-contrib/providers/flagd/pkg/service/grpc"
+	HTTPService "github.com/open-feature/golang-sdk-contrib/providers/flagd/pkg/service/http"
+	of "github.com/open-feature/golang-sdk/pkg/openfeature"
+)
+
+type Provider struct {
+	service     service.IService
+	port        int32
+	host        string
+	serviceName ServiceType
+	protocol    string
+}
+
+type ServiceType int
+
+const (
+	HTTP ServiceType = iota
+	GRPC
+)
+
+type ProviderOption func(*Provider)
+
+func NewProvider(opts ...ProviderOption) *Provider {
+	provider := &Provider{
+		serviceName: HTTP,
+		port:        8080,
+		host:        "localhost",
+		protocol:    "http",
+	}
+	for _, opt := range opts {
+		opt(provider)
+	}
+	if provider.serviceName == GRPC {
+		provider.service = GRPCService.NewGRPCService(
+			GRPCService.WithPort(provider.port),
+			GRPCService.WithHost(provider.host),
+		)
+	} else {
+		provider.service = HTTPService.NewHTTPService(
+			HTTPService.WithPort(provider.port),
+			HTTPService.WithHost(provider.host),
+			HTTPService.WithProtocol(provider.protocol),
+		)
+	}
+	return provider
+}
+
+// WithHost specifies the port of the flagd server. Defaults to 8080
+func WithPort(port int32) ProviderOption {
+	return func(p *Provider) {
+		p.port = port
+	}
+}
+
+// WithHost specifies the host name of the flagd server. Defaults to localhost
+func WithHost(host string) ProviderOption {
+	return func(p *Provider) {
+		p.host = host
+	}
+}
+
+// WithService specifies the type of the service. service should be one of "http" or "grpc", if not the default "http" will be used
+func WithService(service ServiceType) ProviderOption {
+	return func(p *Provider) {
+		p.serviceName = service
+	}
+}
+
+// WithProtocol specifies the protocol used by the http service. Should be one of "http" or "https", if not the default "http" will be used, https is not currently supported
+func WithProtocol(protocol string) ProviderOption {
```

Comment:
nitpick: I think you could introduce a `Protocol` enum here, like you did for `ServiceType`. Please also adapt the comments speaking about what can be passed as a value.... (same for WithService(...))

Reply:
updated in latest commit

Reply:
Mhm...., I agree that it doesn't look nice. Looking at it, does it make sense to cover all of this by just providing a `WithProtocol` option ? :thinking: 
```
NewProvider(WithProtocol(flagd.HTTP))) or
NewProvider(WithProtocol(flagd.HTTPS))) or
NewProvider(WithProtocol(flagd.GRPC))
```

Reply:
could also update the enum to be evaluated at the top level?

```
type Provider struct {
	service     service.IService
	port        int32
	host        string
	serviceName ServiceType
	protocol    string
}

func WithProtocol(protocol HTTPService.ProtocolType) ProviderOption {
	if protocol == HTTPS {
		return func(p *Provider) {
			p.protocol = "https"
		}
	}
	return function(p *Provider) {
		p.protocol = "http"
	}
}
```

so the HTTP service is passed the correct string value and is unaware of the provider level enum

Reply:
hm yeah i like the idea :) However, one thing which bugs me is that, `protocol` is a field of `Provider` whereas it is actually only applicable to HTTP based service implementation. (correct?). Then, for me it would indeed make sense to just remove this field `protocol` from the Provider level and distinguish between three different types of services e.g. using `NewProvider(WithService(HTTP|HTTPS|GRPC))` on the provider level, and get rid of `WithProtocol`. Or do you have different view on this? Maybe additional oppinions would help. @skyerus ? :-)

Reply:
that's a good point, I think it will also be the cleanest implementation, you've convinced me!